AutoGen Blog (#99)
mlflow/mlflow-website: website/blog/2024-08-29-autogen-pyfunc/index.md

## 1 - Setup

First, let's install the required dependencies. Note that pyautogen requires `python>=3.9`.

### 1.1 - Environment Setup

```shell
%pip install pyautogen mlflow -U -q
```

We must also get API credentials to use an LLM. For this tutorial, we'll be using OpenAI. Note that a great way to securely pass tokens to your interactive python environment is via the [getpass](https://docs.python.org/3/library/getpass.html) package.

```python
import os
from getpass import getpass

os.environ["OPENAI_API_KEY"] = getpass("OPENAI_API_KEY:")

assert os.getenv("OPENAI_API_KEY"), "Please set an OPENAI_API_KEY environment variable."
```

> Great! We've authenticated and are ready to start building an agent framework.

Reviewer: ```suggestion
Great! We've configured our authentication configuration and are ready to start building an agent framework.
```

    pil_img = get_pil_image(img_data)
    return pil_img

class DALLEAgent(ConversableAgent):
    def __init__(self, name, llm_config: dict, **kwargs):
        super().__init__(name, llm_config=llm_config, **kwargs)

        api_key = os.getenv("OPENAI_API_KEY")
        self._dalle_client = OpenAI(api_key=api_key)
        self.register_reply([Agent, None], DALLEAgent.generate_dalle_reply)

    def send(
        self,
        message: Union[Dict, str],
        recipient: Agent,
        request_reply: Optional[bool] = None,
        silent: Optional[bool] = False,
    ):
        # override and always "silent" the send out message;
        # otherwise, the print log would be super long!
        super().send(message, recipient, request_reply, silent=True)

    def generate_dalle_reply(self, messages: Optional[List[Dict]], sender: "Agent", config):
        """Generate a reply using OpenAI DALLE call."""
        client = self._dalle_client if config is None else config
        if client is None:
            return False, None
        if messages is None:
            messages = self._oai_messages[sender]

        prompt = messages[-1]["content"]

>         # TODO: integrate with autogen.oai. For instance, with caching for the API call

Reviewer: Can we remove the TODO messages?

        img_data = dalle_call(
            client=client,
            model="dall-e-3",
            prompt=prompt,

>             size="1024x1024",  # TODO: the size should be flexible, deciding landscape, square, or portrait mode.

Reviewer: let's remove this. The values that can be specified are an enumeration. No need to complicate the example with that. hard-coding the spec is fine.

            name="Critics",
            system_message=f"""You need to improve the prompt of the figures you saw.
{CRITIC_PROMPT}
Reply with the following format:

CRITICS: the image needs to improve...
PROMPT: here is the updated prompt!

""",
            llm_config={"max_tokens": 1000, "model": "gpt-4o"},
            human_input_mode="NEVER",
            max_consecutive_auto_reply=3,
        )

        self.dalle = DALLEAgent(
            name="Dalle", llm_config={"model": "dalle"}, max_consecutive_auto_reply=0
        )

        # Data flow begins
        self.send(message=img_prompt, recipient=self.dalle, request_reply=True)
        img = extract_img(self.dalle)
        plt.imshow(img)
        plt.axis("off")  # Turn off axis numbers
        plt.show()
        print("Image PLOTTED")

        for i in range(self._n_iters):
            # Downsample the image s.t. GPT-4V can take
            img = extract_img(self.dalle)
            smaller_image = img.resize((128, 128), Image.Resampling.LANCZOS)
            smaller_image.save("result.png")

            self.msg_to_critics = f"""Here is the prompt: {img_prompt}.
            Here is the figure <img result.png>.

>             Now, critic and create a prompt so that DALLE can give me a better image.

Reviewer: ```suggestion
            Now, critique and create a prompt so that DALLE can give me a better image.
```

            llm_config={"model":"gpt-4"},
        )

    user_proxy = UserProxyAgent(name="User", human_input_mode="NEVER", max_consecutive_auto_reply=0, code_execution_config={
                "work_dir": "output", # Location where code will be written
                "use_docker": False # Use local jupyter execution environment instead of docker
            })

    return user_proxy.initiate_chat(
        creator, message=model_input
    )
mlflow.models.set_model(CatifyPyfunc())
```

At the end of this step, you should have a Python file that has both code snippets. The name is up to the user, but for this blog we will use "catify_model.py".

## 3 Use Our Agent Framework

We are now positioned to leverage MLflow to interact with our powerful "catify" agent.

### 3.1 - Log and Load

First, let's demonstrate the standard user journey of logging model to MLflow's tracking server. We will then load it back and perform inference.

```python
import mlflow
mlflow.autogen.autolog() # Enable logging of traces

with mlflow.start_run() as run:
    mlflow.pyfunc.log_model(
        artifact_path="autogen_pyfunc",
        python_model="catify_model.py", # Our model from code python file

> 

Reviewer: It might be worthwhile to log a JSON artifact to the run that has 2 prompts for 'testing' and using the `log_image` API for storing the resulting generated images just to showcase this functionality that exists within MLflow. The dict that is logged can be a mapping of the index of input to both the prompt and the logged image name. 
No need to re-generate the images for the examples here - they're great - just update the logging function here to iterate over a list of test prompts and construct those data structures / call the model to log the images.

Author: Autogen cache for the win! Great call

2. Create a private utility to increment our metadata and log and images with [mlflow.log_image](https://mlflow.org/docs/latest/python_api/mlflow.html?highlight=log_image#mlflow.log_image).
3. Call the above utility after new images have been generated.
4. Finally, log our metadata object as JSON with [mlflow.log_dict](https://mlflow.org/docs/latest/python_api/mlflow.html?highlight=log_image#mlflow.log_dict).

```diff
+ import uuid

class CatifyWithDalle(AssistantAgent):
    def __init__(self, n_iters=2, **kwargs):
        """
        Initializes a CatifyWithDalle instance.

        This agent facilitates the creation of visualizations through a collaborative effort among
        its child agents: dalle and critics.

        Parameters:
            - n_iters (int, optional): The number of "improvement" iterations to run. Defaults to 2.
            - **kwargs: keyword arguments for the parent AssistantAgent.
        """
        super().__init__(**kwargs)
        self.register_reply([Agent, None], reply_func=CatifyWithDalle._reply_user, position=0)
        self._n_iters = n_iters
+       self.dict_to_log = {}

> +   def _log_image_and_append_to_dict(self, img: Image, img_prompt: str, image_index: int)-> None:
> +      # Generate a unique ID
> +     _id = str(uuid.uuid1())
> 
> +       # Append to class variable to log once at the end of all inference
> +       self.dict_to_log[_id] = {"prompt": img_prompt, "index": image_index}
> 
> +       # Log image to MLflow tracking server
> +       mlflow.log_image(img, f"{_id}.png")

Reviewer: ```suggestion
+    def _log_image_and_append_to_dict(self, img: Image, img_prompt: str, image_index: int)-> None:
+        # Generate a unique ID
+        _id = str(uuid.uuid1())

+        # Append to class variable to log once at the end of all inference
+        self.dict_to_log[_id] = {"prompt": img_prompt, "index": image_index}

+        # Log image to MLflow tracking server
+        mlflow.log_image(img, f"{_id}.png")
```
